### PR TITLE
feat(core): 11343 cancel all request to api from resource atom on new request by default

### LINKED ISSUE #85: Core changes - better api for createResourseAtom proposal
Now we have next createResourceAtom signature

```ts
type FetcherFunc<P, T> = (params?: P | null) => Promise<T>;
type FetcherProcessor<T> = () => Promise<T>;
type FetcherCanceller<P> = (context: ResourceCtx<P>) => void;
type FetcherFabric<P, T> = (params?: P | null) => {
  processor: FetcherProcessor<T>;
  canceller?: FetcherCanceller<P>;
  allowCancel?: boolean;
};

export function createResourceAtom<P, T>(
  fetcher: FetcherFunc<P, T> | FetcherFabric<P, T>,
  paramsAtom?: Atom<P> | ResourceAtom<any, any> | null,
  name = `Resource-${resourceAtomIndex++}`,
  isLazy = false
): ResourceAtomType<P, T>;
```

## Example of usage:

Case - simplest one: resource with no parameters that fetched once after created

```ts
const weatherResource = createResourceAtom<null, WeatherState>(
  () => {
    return apiClient.get<WeatherState[]>('/weather');
  },
  null,
  'weatherResource'
);
```

Case - resource with parameters: will be requested every time after weatherUnits changed

```ts
const weatherResource = createResourceAtom<WeatherUnits, WeatherState>(
  (weatherUnits) => {
    return apiClient.get<WeatherState[]>(`/weather?units=${weatherUnits}`);
  },
  weatherUnitsAtom,
  'weatherResource'
);
```

Case - lazy resource with parameters: will be requested if have subscribers when weatherUnits changed

```ts
const weatherResource = createResourceAtom<WeatherUnits, WeatherState>(
  (weatherUnits) => {
    return apiClient.get<WeatherState[]>(`/weather?units=${weatherUnits}`);
  },
  weatherUnitsAtom,
  'weatherResource',
  true
);
```

Case - lazy resource that can be canceled

```ts
const weatherResource = createResourceAtom<WeatherUnits, WeatherState>(
  (weatherUnits) => {
    const abortController = new AbortController();

    async function processor() {
      try {
        return await apiClient.get<WeatherState[]>(`/weather?units=${weatherUnits}`, true, {
          signal: abortController.signal,
          errorsConfig: { dontShowErrors: true },
        });
      } catch (e) {
        if (isApiError(e) && e.problem.kind === 'canceled') {
          return null;
        }
        throw e;
      }
    }

    function canceller() {
      try {
        abortController.abort();
      } catch (e) {
        console.warn('Cannot abort previous request!', e);
      }
    }

    return { processor, canceller };
  },
  weatherUnitsAtom,
  'weatherResource',
  true
);
```

## Proposal

1. Pass dependencies as first argument.
   In complex function it's will be much easer to read

```ts
const weatherResource = createResourceAtom<WeatherUnits, WeatherState>(
  weatherUnitsAtom, // Now it closer to fetcher function arguments
  (weatherUnits) => {
    return apiClient.get<WeatherState[]>(`/weather?units=${weatherUnits}`);
  },
  'weatherResource',
  true
);
```

2. Last argument should be options object - this allow us understand what this 'true' means

```ts
const weatherResource = createResourceAtom<WeatherUnits, WeatherState>(
  weatherUnitsAtom,
  (weatherUnits) => {
    return apiClient.get<WeatherState[]>(`/weather?units=${weatherUnits}`);
  },
  { name: 'weatherResource', lazy: true }
);
```

3. Pass abort controller to feather, for cancel request automatically.
- and remove error throwing after request aborted

```ts
const weatherResource = createResourceAtom<WeatherUnits, WeatherState>(
  weatherUnitsAtom,
  (weatherUnits, abortController) => {
     return await apiClient.get<WeatherState[]>(`/weather?units=${weatherUnits}`, signal: abortController.signal);
  },
  { name: 'weatherResource', lazy: true },
);
```

### PR DESCRIPTION
**Pull request by Akiyamka**
closes #85

### CODE REVIEW

File: src/components/Analytics/AnalyticsPanelHeader/AnalyticsPanelHeader.tsx [hunk 1]
```diff
@@ -26,7 +28,9 @@ function PanelHeading({ event }: PanelHeadingProps) {
 }
 
 interface AnalyticsPanelHeaderParams {
-  resourceAtom: ResourceAtomType<any, any>;
+  resourceAtom:
+    | Atom<AsyncAtomState<FocusedGeometry | null, AnalyticsData[] | null>>
+    | Atom<AsyncAtomState<FocusedGeometry | null, AdvancedAnalyticsData[] | null>>;
```
Comment (IlyaIzr) [inline src/components/Analytics/AnalyticsPanelHeader/AnalyticsPanelHeader.tsx]:
```suggestion
    Atom<AsyncAtomState<FocusedGeometry | null, AnalyticsData[] | null>>
    | Atom<AsyncAtomState<FocusedGeometry | null, AdvancedAnalyticsData[] | null>>;
```

Comment (IlyaIzr) [reply src/components/Analytics/AnalyticsPanelHeader/AnalyticsPanelHeader.tsx]: Anyway this is really hard to digest 😄

File: src/core/auth/atoms/userResource.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import { apiClient } from '~core/apiClientInstance';
+import { createAsyncAtom } from '~utils/atoms/createAsyncAtom';
 import { createAtom, createResourceAtom } from '~utils/atoms';
```
Comment (IlyaIzr) [inline src/core/auth/atoms/userResource.ts]:
```suggestion
import { createAtom } from '~utils/atoms';
```

File: src/core/auth/atoms/userResource.ts [hunk 3]
```diff
@@ -125,5 +109,94 @@ export const userResourceAtom = createResourceAtom<
...
+//         }
+//       } catch (e) {}
+//     }
+
+//     const udm = new UserDataModel({ features, feeds });
+//     return udm;
+//   },
+//   'userResourceAtom',
+//   userResourceRequestParamsAtom,
+// );
```
Comment (IlyaIzr) [inline src/core/auth/atoms/userResource.ts]: either delete it or leave a comment why to keep it

File: src/utils/atoms/createResourceAtom/readme.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
+/\*\*
+
+- You can chain this resources!
```
Comment (propakov) [inline src/utils/atoms/createResourceAtom/readme.md]:
```suggestion
- You can chain these resources!
```

File: src/utils/atoms/createResourceAtom/readme.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        return await apiClient.get<WeatherState[]>(
+          `/weather?units=${weatherUnits}`,
+          true,
+          {
+            signal: abortController.signal,
+            errorsConfig: { dontShowErrors: true },
+          },
+        );
+      } catch (e) {
+        if (isApiError(e) && e.problem.kind === 'canceled') {
```
Comment (propakov) [inline src/utils/atoms/createResourceAtom/readme.md:84]: maybe also add helper function like ?
`const isCanceled = (e) => isApiError(e) && e.problem.kind === 'canceled'`
I've just found this code is repeated everywhere

File: src/utils/atoms/createResourceAtom/readme.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+const weatherResource = createResourceAtom<WeatherUnits, WeatherState>(
+  weatherUnitsAtom,
+  (weatherUnits) => {
+    return apiClient.get<WeatherState[]>(`/weather?units=${weatherUnits}`);
+  },
+  { name: 'weatherResource', lazy: true },
+);
+```
+
+3. Pass abort controller to feather, for cancel request automatically.
```
Comment (propakov) [inline src/utils/atoms/createResourceAtom/readme.md]: `to feather` feels like typo 🤔

File: src/utils/atoms/createResourceAtom/readme.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
+/\*\*
+
+- You can chain this resources!
+- The demo:
+- https://codesandbox.io/s/reatom-resource-atom-complex-qwi3h
```
Comment (propakov) [inline src/utils/atoms/createResourceAtom/readme.md:4]: would be cool to have at least some short  1-line description of what's going on on demo

File: src/utils/atoms/createResourceAtom/createResourceAtom.ts [hunk 8]
```diff
@@ -226,51 +226,44 @@ export function createResourceAtom<P, T>(
...
-              }
-              if ('data' in newParams) {
-                dispatch(resourceFetcherAtom.request(newParams.data));
-                return;
-              }
+    const autoFetchAtom = createAtom({ paramsAtom }, ({ onChange, schedule }) => {
+      onChange('paramsAtom', (newParams, prevParams) => {
+        schedule((dispatch) => {
+          if (isObject(newParams)) {
+            // Check states than we can be escalated
```
Comment (propakov) [inline src/utils/atoms/createResourceAtom/createResourceAtom.ts]:
```suggestion
            // Check states than can be escalated
```

File: src/utils/atoms/createAsyncAtom/createAsyncAtom.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+            schedule((dispatch) => dispatch(create('request', depsAtomState as any)));
+          }
+        });
+      } else {
+        if (options.auto && !newState.dirty) {
+          schedule((dispatch) => dispatch(create('request', null as any)));
+        }
+      }
+
+      // return notDeepMemo(newState);
```
Comment (propakov) [inline src/utils/atoms/createAsyncAtom/createAsyncAtom.ts]: it looks like forgotten unremoved line

File: src/core/shared_state/defaultLayers.test.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  export interface TestContext {
+    store: Store;
+  }
+}
+
+beforeEach(async (context) => {
+  context.store = createStore();
+});
+
+test.skip('Default layers', async ({ store }) => {
```
Comment (propakov) [inline src/core/shared_state/defaultLayers.test.ts:23]: so we skip this new test, right? if yes, why?

File: src/utils/atoms/createResourceAtom/readme.md (hunk 5 above)
Comment (propakov) [reply src/utils/atoms/createResourceAtom/readme.md:84]: is see we have` isAbortError`

File: src/core/auth/atoms/userResource.ts (hunk 3 above)
Comment (Akiyamka) [reply src/core/auth/atoms/userResource.ts]: Oops, forgot to remove this

File: src/utils/atoms/createAsyncAtom/createAsyncAtom.ts [hunk 11]
```diff
@@ -0,0 +1,254 @@
+import { isObject } from '@reatom/core';
+import { memo } from '@reatom/core/experiments';
+import { createAtom } from '~utils/atoms/createPrimitives';
+import { store } from '~core/store/store';
+import { isErrorWithMessage } from '~utils/common';
+import { abortable, ABORT_ERROR_MESSAGE, isAbortError } from './abort-error';
+import { isAtomLike } from './is-atom-like';
+import type { AsyncAtomOptions, AsyncAtomState, Fetcher, AsyncAtomDeps } from './types';
+import type { Atom, AtomBinded, AtomSelfBinded, AtomState } from '@reatom/core';
```
Comment (propakov) [inline src/utils/atoms/createAsyncAtom/createAsyncAtom.ts]:
```suggestion
import type { AtomBinded, AtomSelfBinded, AtomState } from '@reatom/core';
```

File: src/utils/atoms/createResourceAtom/readme.md (hunk 5 above)
Comment (Akiyamka) [reply src/utils/atoms/createResourceAtom/readme.md:84]: Yeah, it's readme for legacy createResouceAtom that i live untouched, but thankfully for you note I remember that I forgot to write readme for new createAsyncAtom!

File: src/core/shared_state/defaultLayers.test.ts (hunk 10 above)
Comment (Akiyamka) [reply src/core/shared_state/defaultLayers.test.ts:23]: Oh, thank you, i forget to create task for this, will add it

**Comment by propakov** [comment]
can't see geometry loaded on the map, but it works on main branch, maybe it's broken as network requests are without errors
Screenshot 2022-10-06 at 15 18 05

File: src/utils/atoms/createResourceAtom/readme.md (hunk 7 above)
Comment (Akiyamka) [reply src/utils/atoms/createResourceAtom/readme.md:4]: `You can chain resources - they inherit state from parent`
Something like this?

Comment (propakov) [reply src/utils/atoms/createResourceAtom/readme.md:4]:
```
You can chain resources - they inherit state from parent
The demo:
- https://codesandbox.io/s/reatom-resource-atom-complex-qwi3h
On demo you see a list of atoms' states, every atom is the child of previous. Cancelling atom will cancel all its children. 
```

File: src/utils/atoms/createResourceAtom/readme.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+```
+
+## Proposal
+
+1. Pass dependencies as first argument.
+   In complex function it's will be much easer to read
+
+```ts
+const weatherResource = createResourceAtom<WeatherUnits, WeatherState>(
+  weatherUnitsAtom, // Now it closer to feather function arguments
```
Comment (propakov) [inline src/utils/atoms/createResourceAtom/readme.md]: I believe you ment `fetcher`  🤔

**Review by propakov — APPROVED** [review]

**Pull request merged**